src/app/InteractionModelEngine.cpp

```diff
@@ -530,5 +530,125 @@ bool InteractionModelEngine::IsOverlappedAttributePath(ClusterInfo & aAttributeP
     return false;
 }
 
+void InteractionModelEngine::DispatchCommand(CommandHandler & apCommandObj, const ConcreteCommandPath & aCommandPath,
+                                             TLV::TLVReader & apPayload)
+{
+    CommandHandlerInterface * handler = FindCommandHandler(aCommandPath.mEndpointId, aCommandPath.mClusterId);
+
+    if (handler)
+    {
+        CommandHandlerInterface::HandlerContext context(apCommandObj, aCommandPath, apPayload);
+        handler->InvokeCommand(context);
+
+        //
+        // If the command was handled, don't proceed any further and return successfully.
+        //
+        if (context.mCommandHandled)
+        {
+            return;
+        }
+    }
+
+    DispatchSingleClusterCommand(aCommandPath, apPayload, &apCommandObj);
+}
+
+bool InteractionModelEngine::CommandExists(const ConcreteCommandPath & aCommandPath)
+{
+    CommandHandlerInterface * handler = FindCommandHandler(aCommandPath.mEndpointId, aCommandPath.mClusterId);
+    if (handler)
+    {
+        return true;
+    }
+
+    //
+    // If we couldn't find a matching handler, default to querying the application using the following method
+    // to deduce if a command exists.
+    //
+    return ServerClusterCommandExists(aCommandPath);
+}
+
+CHIP_ERROR InteractionModelEngine::RegisterCommandHandler(CommandHandlerInterface * handler)
+{
+    VerifyOrReturnError(handler != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    for (auto * cur = mpCommandHandlerList; cur; cur = cur->GetNext())
+    {
+        if (cur->Matches(*handler))
+        {
+            ChipLogError(InteractionModel, "Duplicate command handler registration failed");
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+    }
+
+    handler->SetNext(mpCommandHandlerList);
+    mpCommandHandlerList = handler;
+
+    return CHIP_NO_ERROR;
+}
+
+void InteractionModelEngine::DeregisterCommandHandlers(EndpointId endpointId)
+{
+    CommandHandlerInterface * prev = nullptr;
+
+    for (auto * cur = mpCommandHandlerList; cur; cur = cur->GetNext())
+    {
+        if (cur->MatchesExactly(endpointId))
```

Comment:
It feels like it will be slightly confusing that this will not actually deregister all handlers if some of them have endpoint as a wildcard:

```
foreach ep in endpoints:
    DeregisterCommandHandlers(ep)
```

Not sure if there's a better way though.  I suppose this does support an important use case, which I think is dynamically adding and removing endpoints without affecting other endpoints.

Reply:
The idea is to unregister endpoint-specific command handlers that are attached to the endpoint that is going away, but leaving the "applies to all endpoints" ones, because those should not be affected by changes in the set of endpoint, right?